lib/sage.ex

```diff
@@ -437,6 +447,57 @@ defmodule Sage do
   def run_async(sage, name, transaction, compensation, opts \\ []),
     do: add_stage(sage, name, build_operation!(:run_async, transaction, compensation, opts))
 
+  @doc """
+  For a given Sage S with transactions :t1 -> :t2 -> :t3, a call to `interleave(S, :name, f)`
+  will yield a saga with transactions :t1 -> :name_1 -> :t2 -> :name_2 -> :t3 -> :name_3.
+
+  This can be useful if you are trying to do a long computation and want to do something with
+  the intermediate results, such as logging or persistence.
+  
+  Note: 
+  - This isn't strict interleaving because a transaction is still appended at the end.
+  - Calling this function twice with the same name will give a `Sage.DuplicateStageError`, but
+  calling the function twice with a different name will compound its effects.
+  - Calling this function before the end of your saga definition will mean any stages you add after the
+  call to `interleave` will not have the intermediate stages added after them
+  """
+  @spec interleave(
+          sage :: t(),
+          name :: stage_name(),
+          intermediate_transaction :: intermediate_transaction(),
+          compensation :: compensation()
+        ) :: t()
+  def interleave(sage, name, intermediate_transaction, compensation \\ :noop)
+      when is_intermediate_transaction(intermediate_transaction) and is_compensation(compensation) do
+    sage.stages
+    |> Enum.reverse()
+    |> Enum.with_index(1)
+    |> Enum.flat_map(fn {stage, index} ->
+      name = String.to_atom("#{name}_#{index}")
+      {stage_name, _} = stage
+
+      transaction =
+        case intermediate_transaction do
+          {m, f, a} ->
+            {m, f, [stage_name | a]}
+
+          _ ->
+            &intermediate_transaction.(&1, &2, stage_name)
+        end
+
+      [stage, {name, build_operation!(:run, transaction, compensation)}]
```

Comment:
What if we use a tuple instead of a binary for the name?
```suggestion
      [stage, {{:interleave, name, stage_name}, build_operation!(:run, transaction, compensation)}]
```

Reply:
I'm not opposed, but I'm not sure I understand the intent behind the suggestion. Could you help me understand what merits that would have?

Reply:
@apreifsteck Sage stage name is a `term()` so it can be anything, which means that the `name = String.to_atom("#{name}_#{index}")` can lead to an exception if `name` contains anything that can't be cast to string. Having a tuple there has a few other benefits:
1. We guarantee that `interleave` has it's namespace and its name won't collide with ordinary stage names;
2. We can also extend the `Inspect` protocol to print interleaves separately and not as regular stages, thus reducing the output and increasing the readability.

Reply:
Done